props/saber_BC_buttons.h

```diff
@@ -386,31 +1134,317 @@ EFFECT(volmax);     // for maximum volume reached
-            triggered_ = false;  // Reset the timer flag
-            return true;  // Timer has elapsed
-        }
-        return false;  // Timer is still running
+  void say() override {
+    if (last_volume_ < current_volume_) {
+      mode::getSL<SPEC>()->SayVolumeUp();
+    } else if (last_volume_ > current_volume_) {
+      mode::getSL<SPEC>()->SayVolumeDown();
+    }
+    // Get the actual updated current volume just in case
+     float volume = dynamic_mixer.get_volume();
+    percentage_ = round((volume / max_volume_) * 10) * 10;
+    if (percentage_ <= 10) {
+      QuickMinVolume();
+    } else if (percentage_ >=100) {
+     QuickMaxVolume();
+    }
+    last_volume_ = current_volume_;
+  }
+
+  void next() override {
+    if (current_volume_ < max_volume_) {
+      current_volume_ += max_volume_ * 0.10;
+    }
+    dynamic_mixer.set_volume(current_volume_);
+  }
+
+  void prev() override {
+    if (current_volume_ > min_volume_) {
+      current_volume_ -= max_volume_ * 0.10;
+    }
+    dynamic_mixer.set_volume(current_volume_);
+  }
+
+  void QuickMaxVolume() {
+    current_volume_ = max_volume_;
+    dynamic_mixer.set_volume(current_volume_);
+    PVLOG_NORMAL << "** Maximum Volume\n";
+    mode::getSL<SPEC>()->SayMaximumVolume();
+    last_volume_ = current_volume_;
+  }
+
+  void QuickMinVolume() {
+    current_volume_ = min_volume_;
+    dynamic_mixer.set_volume(current_volume_);
+    PVLOG_NORMAL << "** Minimum Volume\n";
+    mode::getSL<SPEC>()->SayMinimumVolume();
+    last_volume_ = current_volume_;
+  }
+
+  void update() override {
+    SaberBase::DoEffect(EFFECT_VOLUME_LEVEL, 0);
+    // Overridden to substitute the tick sound
+    this->say_time_ = Cyclint<uint32_t>(millis()) + 1;
+    if (!this->say_time_) this->say_time_ += 1;
+    this->fadeout(SaberBase::sound_length);
+  }
+
+  void select() override {
+    cancelling_ = false;
+    mode::getSL<SPEC>()->SaySave();
+    SPEC::SteppedMode::select();
+  }
+
+  void exit() override {
+    PVLOG_NORMAL << "** Exit Volume Menu\n";
+    cancelling_ = true;
+    SPEC::MenuBase::exit();
+  }
+
+  bool mode_Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, 0)) {
+      // Custom button controls for BCVolumeMode
+      case EVENTID(BUTTON_POWER, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMaxVolume();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_SECOND_HELD_MEDIUM, 0):
+      case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMinVolume();
+        return true;
     }
+    // Use the select and exit controls from SelectCancelMode
+    return SPEC::SelectCancelMode::mode_Event2(button, event, modifiers);
+  }
+  bool cancelling_ = false;
+};
+
+#ifdef DYNAMIC_BLADE_LENGTH
+
+int selected_blade_;
+
+#ifdef BC_DUAL_BLADES
+template<class SPEC>
+struct BCSelectBladeMode : public SPEC::MenuBase {
+  int current_blade_ = 1; // Start with the first blade
```

Comment:
why not use pos_ +1 instead?

Reply:
I don't know? I thought current_blade_ was very clear as to what it is.
Also, pos_ is not in scope.

Reply:
Because C++ is weird, you probably have to use this->pos_ to get to it.
It's not about clarity, it's the fact that MenuBase already manages a menu position for you, so I think you should use it, as it makes for less code.

If you're not going to use it, you probably want SteppedMode again...

Reply:
> use this->pos_

Would you be ok with a helper function to help readability? In order to account for the offset, things can get kind of unruly like `this->pos_ = (this->pos_ + size() - 1) % size();`
I could do something like
```cpp
  int current_blade() const {
    return (this->pos_ % size()) + 1;
  }
  ```
and then clean things up by just calling that?
Then it's still using pos_ as requested.

Reply:
this->pos_ should always be between 0 and size() -1, so the % doesn't help.
A helper function  can be nice though, as it gives a better name than "pos_"

Reply:
but i tried that and it cycled through blade 0-3 not 1-4.

Reply:
> MenuBase already manages a menu position for you, so I think you should use it, as it makes for less code.

I'm trying to abide by this, but I don't see how to "do" anything when prev() or next() happens if I can't override them, and if they are overridden, then it's just new, replacement code anyway, in which case enforcing using pos_ seems weird.
If not using pos_ enforcing inheriting from SteppedMode is weird because I don't need the methods it uses.
I'm in a catch 22 it seems.

Reply:
What do you want to do in prev() / next()?

> If not using pos_ enforcing inheriting from SteppedMode is weird because I don't need the methods it uses.

MenuBase is just SteppedMode, but with an added size and position, if you don't want the size and position, then why inherit from MenuBase?

If you don't need the methods from SteppedMode, then maybe you should inherit from SteppedModeBase?

> I'm in a catch 22 it seems.

I don't understand the catch 22 here.

Reply:
> then maybe you should inherit from SteppedModeBase?

oh. I hadn't seen you just made this.

If not inheriting from MenuBase, the modolo is needed, right? 


> What do you want to do in prev() / next()?

Stop the current blade highlighting, increment/decrement the blade #, start highlighting the new blade.